File: webapp/src/transport/csil/rpc.ts
// CSIL-RPC transport — request/response/push envelopes — see `csil-rpc-transport.md`.

import type { Cbor } from "./cbor.ts";
import type { FrameCarrier } from "./carrier.ts";
import {
  Status,
  TransportError,
  VERSION,
  canonMap,
  checkVersion,
  decodeValue,
  encodeValue,
  getInt,
  getText,
  getTextOpt,
  getUint,
  getUintOpt,
  intValue,
  mapGet,
  statusIsOk,
  tag24,
  textValue,
  untag24,
} from "./conventions.ts";

// A CSIL-RPC request (client → server). `payload` is the opaque CBOR(request type)
// bytes; the transport wraps it in tag 24 on the wire and never inspects it.
export class RpcRequest {
  service: string;
  op: string;
  id?: number;
  payload: Uint8Array;
  auth?: string;

  constructor(service: string, op: string, payload: Uint8Array) {
    this.service = service;
    this.op = op;
    this.payload = payload;
  }

  withId(id: number): this {
    this.id = id;
    return this;
  }

  withAuth(auth: string): this {
    this.auth = auth;
    return this;
  }

  encode(): Uint8Array {
    const entries: Array<readonly [string, Cbor]> = [
      ["v", intValue(VERSION)],
      ["service", textValue(this.service)],
      ["op", textValue(this.op)],
      ["payload", tag24(this.payload)],
    ];
    if (this.id !== undefined) entries.push(["id", intValue(this.id)]);
    if (this.auth !== undefined) entries.push(["auth", textValue(this.auth)]);
    return encodeValue(canonMap(entries));
  }

  static decode(bytes: Uint8Array): RpcRequest {
    const v = decodeValue(bytes);
    checkVersion(getUint(v, "v"));
    const payloadField = mapGet(v, "payload");
    if (!payloadField) throw TransportError.malformed("missing 'payload'");
    const req = new RpcRequest(getText(v, "service"), getText(v, "op"), untag24(payloadField));
    req.id = getUintOpt(v, "id");
    req.auth = getTextOpt(v, "auth");
    return req;
  }
}

// A CSIL-RPC response (server → client). `status` is the transport outcome (a
// registry code); `variant` names the declared output-choice arm `payload`
// decodes to when `status` is 0. `payload` is empty when `status` is non-zero.
export class RpcResponse {
  id?: number;
  status: number;
  variant?: string;
  error?: string;
  payload: Uint8Array;

  constructor(status: number, payload: Uint8Array) {
    this.status = status;
    this.payload = payload;
  }

  // A successful (status Ok) typed reply.
  static ok(variant: string, payload: Uint8Array): RpcResponse {
    const r = new RpcResponse(Status.Ok, payload);
    r.variant = variant;
    return r;
  }

  // A transport-level failure (no typed payload).
  static transportError(status: number, message: string): RpcResponse {
    const r = new RpcResponse(status, new Uint8Array(0));
    r.error = message;
    return r;
  }

  withId(id: number | undefined): this {
    this.id = id;
    return this;
  }

  encode(): Uint8Array {
    const entries: Array<readonly [string, Cbor]> = [
      ["v", intValue(VERSION)],
      ["status", intValue(this.status)],
      ["payload", tag24(this.payload)],
    ];
    if (this.id !== undefined) entries.push(["id", intValue(this.id)]);
    if (this.variant !== undefined) entries.push(["variant", textValue(this.variant)]);
    if (this.error !== undefined) entries.push(["error", textValue(this.error)]);
    return encodeValue(canonMap(entries));
  }

  static decode(bytes: Uint8Array): RpcResponse {
    const v = decodeValue(bytes);
    checkVersion(getUint(v, "v"));
    // payload is present but may be an empty byte string on error.
    const payloadField = mapGet(v, "payload");
    const payload = payloadField ? untag24(payloadField) : new Uint8Array(0);
    const r = new RpcResponse(getInt(v, "status"), payload);
    r.id = getUintOpt(v, "id");
    r.variant = getTextOpt(v, "variant");
    r.error = getTextOpt(v, "error");
    return r;
  }

  // Throw `TransportError` for a non-ok response so callers surface transport
  // failures distinctly from application errors (which ride as status-0 variants).
  intoTransportError(): RpcResponse {
    if (statusIsOk(this.status)) return this;
    throw TransportError.status(this.status, this.error);
  }
}

// A CSIL-RPC server push (server → client) for `<-` operations. No id (not a
// reply) and no status (cannot fail in the request/response sense).
export class RpcPush {
  service: string;
  event: string;
  payload: Uint8Array;

  constructor(service: string, event: string, payload: Uint8Array) {
    this.service = service;
    this.event = event;
    this.payload = payload;
  }

  encode(): Uint8Array {
    return encodeValue(
      canonMap([
        ["v", intValue(VERSION)],
        ["service", textValue(this.service)],
        ["event", textValue(this.event)],
        ["payload", tag24(this.payload)],
      ]),
    );
  }

  static decode(bytes: Uint8Array): RpcPush {
    const v = decodeValue(bytes);
    checkVersion(getUint(v, "v"));
    const payloadField = mapGet(v, "payload");
    if (!payloadField) throw TransportError.malformed("missing 'payload'");
    return new RpcPush(getText(v, "service"), getText(v, "event"), untag24(payloadField));
  }
}

// The outcome a server handler returns for one request: a typed reply (variant
// name + encoded payload) on success, or a transport status on failure.
export type HandlerOutcome =
  | { kind: "reply"; variant: string; payload: Uint8Array }
  | { kind: "transport"; status: number; message: string };

export const reply = (variant: string, payload: Uint8Array): HandlerOutcome => ({
  kind: "reply",
  variant,
  payload,
});

export const transport = (status: number, message: string): HandlerOutcome => ({
  kind: "transport",
  status,
  message,
});

// A CSIL-RPC client over a frame carrier. The carrier is injected (bring your own);
// the client owns the envelope and a per-connection monotonic correlation id.
export class RpcClient {
  private nextId = 1;
  private readonly carrier: FrameCarrier;
  private readonly multiplexed: boolean;

  // `multiplexed` true assigns a correlation id to every request (required on WS /
  // pipelined streams); false omits it (strictly one-in-flight carriers).
  constructor(carrier: FrameCarrier, multiplexed: boolean) {
    this.carrier = carrier;
    this.multiplexed = multiplexed;
  }

  // Invoke `service/op` with an encoded request payload, returning the decoded
  // response. A non-zero transport status is thrown as `TransportError`.
  async call(
    service: string,
    op: string,
    payload: Uint8Array,
    auth?: string,
  ): Promise<RpcResponse> {
    const req = new RpcRequest(service, op, payload);
    req.auth = auth;
    if (this.multiplexed) {
      req.id = this.nextId;
      this.nextId += 1;
    }
    await this.carrier.sendFrame(req.encode());
    const frame = await this.carrier.recvFrame();
    if (frame === null) {
      throw TransportError.carrier("connection closed before response");
    }
    return RpcResponse.decode(frame).intoTransportError();
  }
}

// A CSIL-RPC server over a frame carrier. The host supplies a handler mapping
// `(service, op, request-payload)` to an outcome; the generated router is the
// natural implementation of that handler.
export class RpcServer {
  private readonly carrier: FrameCarrier;

  constructor(carrier: FrameCarrier) {
    this.carrier = carrier;
  }

  // Read one request, dispatch it through `handler`, and write the response.
  // Returns `false` at a clean end of stream.
  async serveOne(handler: (req: RpcRequest) => HandlerOutcome | Promise<HandlerOutcome>): Promise<boolean> {
    const frame = await this.carrier.recvFrame();
    if (frame === null) return false;

    let resp: RpcResponse;
    try {
      const req = RpcRequest.decode(frame);
      const outcome = await handler(req);
      resp =
        outcome.kind === "reply"
          ? RpcResponse.ok(outcome.variant, outcome.payload).withId(req.id)
          : RpcResponse.transportError(outcome.status, outcome.message).withId(req.id);
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      resp = RpcResponse.transportError(Status.MalformedEnvelope, message);
    }
    await this.carrier.sendFrame(resp.encode());
    return true;
  }
}

File: api/internal/transport/rpc.go
// CSIL-RPC transport — request/response/push envelopes — see csil-rpc-transport.md.
package transport

// RpcRequest is a CSIL-RPC request (client → server).
type RpcRequest struct {
	Service string
	Op      string
	ID      *uint64
	// Payload is the opaque CBOR(request type) bytes (wrapped in tag 24 on the wire).
	Payload []byte
	Auth    *string
}

// RpcResponse is a CSIL-RPC response (server → client).
type RpcResponse struct {
	ID     *uint64
	Status Status
	// Variant names which declared output-choice arm Payload decodes to (the CSIL type name).
	Variant *string
	Error   *string
	// Payload is the opaque CBOR(output type) bytes; empty when Status is non-zero.
	Payload []byte
}

// RpcPush is a CSIL-RPC server push (server → client) for `<-` operations.
type RpcPush struct {
	Service string
	Event   string
	Payload []byte
}

// NewRpcRequest builds a request with no correlation id and no per-request auth.
func NewRpcRequest(service, op string, payload []byte) RpcRequest {
	return RpcRequest{Service: service, Op: op, Payload: payload}
}

// WithID sets the correlation id (required on multiplexed carriers).
func (r RpcRequest) WithID(id uint64) RpcRequest {
	r.ID = &id
	return r
}

// WithAuth sets the per-request credential for caller-scoped operations.
func (r RpcRequest) WithAuth(auth string) RpcRequest {
	r.Auth = &auth
	return r
}

func (r RpcRequest) Encode() ([]byte, error) {
	entries := []cEntry{
		{cText("v"), cUint(VERSION)},
		{cText("service"), cText(r.Service)},
		{cText("op"), cText(r.Op)},
		{cText("payload"), tag24(r.Payload)},
	}
	if r.ID != nil {
		entries = append(entries, cEntry{cText("id"), cUint(*r.ID)})
	}
	if r.Auth != nil {
		entries = append(entries, cEntry{cText("auth"), cText(*r.Auth)})
	}
	return encodeValue(canonMap(entries)), nil
}

func DecodeRpcRequest(b []byte) (RpcRequest, error) {
	v, _, err := decodeEnvelope(b)
	if err != nil {
		return RpcRequest{}, err
	}
	ver, err := getUint(v, "v")
	if err != nil {
		return RpcRequest{}, err
	}
	if err := checkVersion(ver); err != nil {
		return RpcRequest{}, err
	}
	p, ok := mapGet(v, "payload")
	if !ok {
		return RpcRequest{}, malformed("missing 'payload'")
	}
	payload, err := untag24(p)
	if err != nil {
		return RpcRequest{}, err
	}
	service, err := getText(v, "service")
	if err != nil {
		return RpcRequest{}, err
	}
	op, err := getText(v, "op")
	if err != nil {
		return RpcRequest{}, err
	}
	return RpcRequest{
		Service: service,
		Op:      op,
		ID:      getUintOpt(v, "id"),
		Payload: payload,
		Auth:    getTextOpt(v, "auth"),
	}, nil
}

// NewRpcResponseOk builds a successful (Status ok) typed reply.
func NewRpcResponseOk(variant string, payload []byte) RpcResponse {
	return RpcResponse{Status: StatusOk, Variant: &variant, Payload: payload}
}

// NewRpcResponseTransportError builds a transport-level failure (no typed payload).
func NewRpcResponseTransportError(status Status, message string) RpcResponse {
	return RpcResponse{Status: status, Error: &message, Payload: []byte{}}
}

// WithID sets the echoed correlation id.
func (r RpcResponse) WithID(id *uint64) RpcResponse {
	r.ID = id
	return r
}

func (r RpcResponse) Encode() ([]byte, error) {
	entries := []cEntry{
		{cText("v"), cUint(VERSION)},
		{cText("status"), cInt(r.Status.Code())},
		{cText("payload"), tag24(r.Payload)},
	}
	if r.ID != nil {
		entries = append(entries, cEntry{cText("id"), cUint(*r.ID)})
	}
	if r.Variant != nil {
		entries = append(entries, cEntry{cText("variant"), cText(*r.Variant)})
	}
	if r.Error != nil {
		entries = append(entries, cEntry{cText("error"), cText(*r.Error)})
	}
	return encodeValue(canonMap(entries)), nil
}

func DecodeRpcResponse(b []byte) (RpcResponse, error) {
	v, _, err := decodeEnvelope(b)
	if err != nil {
		return RpcResponse{}, err
	}
	ver, err := getUint(v, "v")
	if err != nil {
		return RpcResponse{}, err
	}
	if err := checkVersion(ver); err != nil {
		return RpcResponse{}, err
	}
	// payload is present but may be an empty byte string on error.
	payload := []byte{}
	if p, ok := mapGet(v, "payload"); ok {
		payload, err = untag24(p)
		if err != nil {
			return RpcResponse{}, err
		}
	}
	status, err := getInt(v, "status")
	if err != nil {
		return RpcResponse{}, err
	}
	return RpcResponse{
		ID:      getUintOpt(v, "id"),
		Status:  StatusFromCode(status),
		Variant: getTextOpt(v, "variant"),
		Error:   getTextOpt(v, "error"),
		Payload: payload,
	}, nil
}

// AsTransportError returns a StatusError for a non-ok response, or nil when the
// response carries a typed reply (status 0). Callers use this after decode to
// surface transport failures distinctly from application errors.
func (r RpcResponse) AsTransportError() error {
	if r.Status.IsOk() {
		return nil
	}
	msg := ""
	if r.Error != nil {
		msg = *r.Error
	}
	return StatusError{Status: r.Status.Name(), Code: r.Status.Code(), Message: msg}
}

// NewRpcPush builds a server-push envelope for a `<-` operation.
func NewRpcPush(service, event string, payload []byte) RpcPush {
	return RpcPush{Service: service, Event: event, Payload: payload}
}

func (p RpcPush) Encode() ([]byte, error) {
	entries := []cEntry{
		{cText("v"), cUint(VERSION)},
		{cText("service"), cText(p.Service)},
		{cText("event"), cText(p.Event)},
		{cText("payload"), tag24(p.Payload)},
	}
	return encodeValue(canonMap(entries)), nil
}

func DecodeRpcPush(b []byte) (RpcPush, error) {
	v, _, err := decodeEnvelope(b)
	if err != nil {
		return RpcPush{}, err
	}
	ver, err := getUint(v, "v")
	if err != nil {
		return RpcPush{}, err
	}
	if err := checkVersion(ver); err != nil {
		return RpcPush{}, err
	}
	p, ok := mapGet(v, "payload")
	if !ok {
		return RpcPush{}, malformed("missing 'payload'")
	}
	payload, err := untag24(p)
	if err != nil {
		return RpcPush{}, err
	}
	service, err := getText(v, "service")
	if err != nil {
		return RpcPush{}, err
	}
	event, err := getText(v, "event")
	if err != nil {
		return RpcPush{}, err
	}
	return RpcPush{Service: service, Event: event, Payload: payload}, nil
}

// RpcClient is a CSIL-RPC client over a frame carrier. The carrier is injected
// (bring your own); the client owns the envelope and a per-connection monotonic
// correlation id.
type RpcClient struct {
	carrier     FrameCarrier
	nextID      uint64
	multiplexed bool
}

// NewRpcClient creates a client. multiplexed true assigns a correlation id to
// every request (required on WS / pipelined streams); false omits it
// (one-in-flight carriers such as HTTP).
func NewRpcClient(carrier FrameCarrier, multiplexed bool) *RpcClient {
	return &RpcClient{carrier: carrier, nextID: 1, multiplexed: multiplexed}
}

// Call invokes service/op with an encoded request payload, returning the decoded
// response. A non-zero transport status is surfaced as a StatusError.
func (c *RpcClient) Call(service, op string, payload []byte, auth *string) (RpcResponse, error) {
	req := NewRpcRequest(service, op, payload)
	req.Auth = auth
	if c.multiplexed {
		id := c.nextID
		req.ID = &id
		c.nextID++
	}
	frame, err := req.Encode()
	if err != nil {
		return RpcResponse{}, err
	}
	if err := c.carrier.SendFrame(frame); err != nil {
		return RpcResponse{}, err
	}
	in, err := c.carrier.RecvFrame()
	if err != nil {
		return RpcResponse{}, err
	}
	if in == nil {
		return RpcResponse{}, ErrCarrier("connection closed before response")
	}
	resp, err := DecodeRpcResponse(in)
	if err != nil {
		return RpcResponse{}, err
	}
	if err := resp.AsTransportError(); err != nil {
		return RpcResponse{}, err
	}
	return resp, nil
}

// Carrier returns the underlying carrier.
func (c *RpcClient) Carrier() FrameCarrier { return c.carrier }

// HandlerOutcome is what a server handler returns for one request: a typed reply
// (variant name + encoded payload) on success, or a transport status on failure.
type HandlerOutcome struct {
	// IsReply distinguishes a typed reply from a transport-status outcome.
	IsReply bool
	Variant string
	Payload []byte
	Status  Status
	Message string
}

// Reply builds a successful handler outcome.
func Reply(variant string, payload []byte) HandlerOutcome {
	return HandlerOutcome{IsReply: true, Variant: variant, Payload: payload}
}

// Transport builds a transport-status handler outcome.
func Transport(status Status, message string) HandlerOutcome {
	return HandlerOutcome{Status: status, Message: message}
}

// RpcServer is a CSIL-RPC server over a frame carrier. The host supplies a handler
// mapping (service, op, request-payload) to an outcome; the generated router is the
// natural implementation of that handler.
type RpcServer struct {
	carrier FrameCarrier
}

func NewRpcServer(carrier FrameCarrier) *RpcServer {
	return &RpcServer{carrier: carrier}
}

// ServeOne reads one request, dispatches it through handler, and writes the
// response. It returns served=false at a clean end of stream.
func (s *RpcServer) ServeOne(handler func(*RpcRequest) HandlerOutcome) (served bool, err error) {
	frame, err := s.carrier.RecvFrame()
	if err != nil {
		return false, err
	}
	if frame == nil {
		return false, nil
	}
	var resp RpcResponse
	if req, derr := DecodeRpcRequest(frame); derr == nil {
		id := req.ID
		outcome := handler(&req)
		if outcome.IsReply {
			resp = NewRpcResponseOk(outcome.Variant, outcome.Payload).WithID(id)
		} else {
			resp = NewRpcResponseTransportError(outcome.Status, outcome.Message).WithID(id)
		}
	} else {
		resp = NewRpcResponseTransportError(StatusMalformedEnvelope, derr.Error())
	}
	out, err := resp.Encode()
	if err != nil {
		return false, err
	}
	if err := s.carrier.SendFrame(out); err != nil {
		return false, err
	}
	return true, nil
}
